src/Correlate.AspNet/Middlewares/CorrelateFeatureNet48.cs

```diff
@@ -0,0 +1,117 @@
+        LoggerMessage.Define<string, string>(LogLevel.Trace, 0xC001, "Setting response header '{HeaderName}' to correlation id '{CorrelationId}'.");
+
+    internal static readonly string RequestActivityKey = $"{typeof(CorrelateFeatureNet48).FullName}, {nameof(RequestActivityKey)}";
+    internal static readonly string CorrelationContextKey = $"{typeof(CorrelateFeatureNet48).FullName}, {nameof(CorrelationContextKey)}";
+
+    private readonly IActivityFactory _activityFactory;
+    private readonly ICorrelationIdFactory _correlationIdFactory;
+    private readonly ILogger _logger;
+    private readonly CorrelateOptionsNet48 _options;
+    
+    public CorrelateFeatureNet48
+    (
+        ILoggerFactory loggerFactory,
+        ICorrelationIdFactory correlationIdFactory,
+        IActivityFactory activityFactory,
+        IOptions<CorrelateOptionsNet48> options
+    )
+    {
+        if (loggerFactory is null)
+        {
+            throw new ArgumentNullException(nameof(loggerFactory));
+        }
+
+        // Even though by contract, we should never have nulls, check anyway. Last thing we want is
+        // for people to experience NRE's in the ASP.NET pipeline and blaming me :)
+        _logger = loggerFactory.CreateLogger(LoggerCategory) ?? throw new InvalidOperationException("The logger factory did not create a logger instance.");
+        _correlationIdFactory = correlationIdFactory ?? throw new ArgumentNullException(nameof(correlationIdFactory));
+        _activityFactory = activityFactory ?? throw new ArgumentNullException(nameof(activityFactory));
+        // ReSharper disable once ConditionalAccessQualifierIsNonNullableAccordingToAPIContract
+        _options = options?.Value ?? throw new ArgumentException("The 'Value' returns null.", nameof(options));
+    }
+    
+    public void StartCorrelating(HttpContextBase httpContext)
+    {
+        (string? _, string correlationId) = GetOrCreateCorrelationHeaderAndId(httpContext);
+
+        IActivity activity = _activityFactory.CreateActivity();
+        CorrelationContext correlationContext = activity.Start(correlationId);
+        // Save the activity so we can clean it up later.
+        httpContext.Items[RequestActivityKey] = activity;
+        httpContext.Items[CorrelationContextKey] = correlationContext;
+    }
+
+    public void StopCorrelating(HttpContextBase httpContext)
+    {
+        // ReSharper disable once InvertIf
+        if (httpContext.Items.TryGetValue(RequestActivityKey, out object? activityObj) &&
+            activityObj is IActivity activity)
+        {
+            (string? responseHeaderName, string correlationId) = GetOrCreateCorrelationHeaderAndId(httpContext);
+            activity.Stop();
+
+            // If already set, ignore.
+            if (httpContext.Response.Headers.TryAdd(responseHeaderName!, correlationId))
+            {
+                LogResponseHeaderAdded(_logger, responseHeaderName!, correlationId, null);
+            }
```

Comment:
`responseHeaderName` may be null when the `IncludeInResponse` option is set to `false`. It would throw here in that case. 

The test `When_response_header_should_not_be_included_and_context_has_started_response_should_not_contain_header` originally covered that but that test does not currently cover this code.

Reply:
This has not been resolved yet.